xwayland/selection/outgoing.c

```diff
@@ -392,6 +392,12 @@ void xwm_handle_selection_request(struct wlr_xwm *xwm,
 		goto fail_notify_requestor;
 	}
 
+	if (req->time <= selection->timestamp) {
+		wlr_log(WLR_DEBUG, "ignoring old request from timestamp %d; expecting > %d",
+				req->time, selection->timestamp);
+		goto fail_notify_requestor;
+	}
+
```

Comment:
From ICCCM:

> The specified owner and selection will be the values that were specified in the SetSelectionOwner request. The owner should compare the timestamp with the period it has owned the selection and, if the time is outside, refuse the SelectionRequest by sending the requestor window a SelectionNotify event with the property set to None (by means of a SendEvent request with an empty event mask).

Reply:
(Note we're not exactly implementing ICCCM's prescribed behavior.)

Reply:
Prior art: https://invent.kde.org/plasma/kwin/-/blob/master/src/xwl/selection.cpp#L237

Though I think check has to be `<` instead of `<=` for safety, due to limited precision in `xcb_timestamp_t`.

Reply:
KWin doesn't have the `XCB_CURRENT_TIME` check and works because https://invent.kde.org/plasma/kwin/-/blob/master/src/xwl/selection.cpp#L107 moves `m_timestamp` back to 0.

Qt has a similar `XCB_CURRENT_TIME` check to what's introduced in this PR, though: https://github.com/qt/qtbase/blob/658b93386eea34398de3ddacab436601e1ab5bbb/src/plugins/platforms/xcb/qxcbclipboard.cpp#L569-L570